Allow Features to declare that they support Rolling upgrades

ryanemerson opened the pull request:
Closes #41022

Metadata generated with the default Features:

```json
{
    "passkeys": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "admin-fine-grained-authz": "{\"enabled\":true,\"version\":2,\"updatePolicy\":\"ROLLING_NO_UPGRADE\"}",
    "dpop": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "impersonation": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "device-flow": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "admin": "{\"enabled\":true,\"version\":2,\"updatePolicy\":\"ROLLING\"}",
    "ciba": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "kerberos": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "login": "{\"enabled\":true,\"version\":2,\"updatePolicy\":\"ROLLING_NO_UPGRADE\"}",
    "token-exchange": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "docker": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "transient-users": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "authorization": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "token-exchange-external-internal": "{\"enabled\":false,\"version\":2,\"updatePolicy\":\"ROLLING\"}",
    "hostname": "{\"enabled\":true,\"version\":2,\"updatePolicy\":\"ROLLING\"}",
    "clusterless": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"SHUTDOWN\"}",
    "recovery-codes": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "rolling-updates": "{\"enabled\":false,\"version\":2,\"updatePolicy\":\"ROLLING_NO_UPGRADE\"}",
    "scripts": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "update-email": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "par": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "multi-site": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"SHUTDOWN\"}",
    "persistent-user-sessions": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"SHUTDOWN\"}",
    "admin-api": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "client-secret-rotation": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "quick-theme": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "user-event-metrics": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "fips": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "ipa-tuura-federation": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "web-authn": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "oid4vc-vci": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "client-types": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "step-up-authentication": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "account-api": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "instagram-broker": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "organization": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "dynamic-scopes": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "declarative-ui": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "client-policies": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "logout-all-sessions": "{\"enabled\":false,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "opentelemetry": "{\"enabled\":true,\"version\":1,\"updatePolicy\":\"ROLLING\"}",
    "account": "{\"enabled\":true,\"version\":3,\"updatePolicy\":\"ROLLING\"}",
    "token-exchange-standard": "{\"enabled\":true,\"version\":2,\"updatePolicy\":\"ROLLING\"}"
}
```

ryanemerson committed:
Allow Features to declare that they support Rolling upgrades

Closes #41022

Signed-off-by: Ryan Emerson <remerson@redhat.com>

pruivo commented on services/src/test/java/org/keycloak/compatibility/FeatureCompatibilityMetadataProviderTest.java:
```diff
@@ -0,0 +1,158 @@
+
+    @ParameterizedTest
+    @MethodSource("shutdownFeatures")
+    public void testFeatureShutdownPolicy(Profile.Feature feature) {
```
Missing both disabled scenario.

pruivo commented on services/src/test/java/org/keycloak/compatibility/FeatureCompatibilityMetadataProviderTest.java:
```diff
@@ -0,0 +1,158 @@
+
+    private static Stream<Arguments> rollingFeatures() {
+        return Arrays.stream(Profile.Feature.values())
+              .filter(f -> f != Profile.Feature.HOSTNAME_V2)
```
Why is `HOSTNAME_V2` skipped?

pruivo commented on services/src/test/java/org/keycloak/compatibility/FeatureCompatibilityMetadataProviderTest.java:
```diff
@@ -0,0 +1,158 @@
+
+    @ParameterizedTest
+    @MethodSource("noUpgradeFeatures")
+    public void testRollingNoUpgradePolicy(Profile.Feature v1, Profile.Feature v2) {
```
missing test switching from v2 enabled to v2 disabled  + v1 enabled

ryanemerson committed:
Pruivo feedback

Signed-off-by: Ryan Emerson <remerson@redhat.com>

ryanemerson commented:
I have addressed all comments.

I have simplified `testRollingNoUpgradePolicy` by just testing login v1/v2, as `ADMIN_FINE_GRAINED_AUTHZ_V2` has a dependency on `AUTHORIZATION` which makes the test logic deviate slightly.

ryanemerson committed:
Fix UpdateCommandDistTest

Signed-off-by: Ryan Emerson <remerson@redhat.com>

ahus1 committed:
Review

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 reviewed (commented):
I've changed ADMIN_FINE_GRAINED_AUTHZ to be an feature that allows rolling update. I rather err on the side of allowing too much rolling :D 
If that is something that core-iam want to change, they can do this in a follow-up PR.

pruivo reviewed (approved):
It's missing a test that adds a feature. I'm fine merging as is if you think the test is an overkill :smiley:

ahus1 commented:
As agreed on the team meeting, please add the test case suggested above to the next PR. Thanks!
